### ISSUE TITLE
Build time increased about 50% in EMSDK 2.0.13 (related to -s USE_SDL=2 -s USE_SDL_TTF=2)

### ISSUE THREAD

**Issue by rsielken** [emscripten-core/emscripten#13676]
In EMSDK 2.0.8, our total build time was about 40 minutes.  When we switched to EMSDK 2.0.13, our total built time jumped to about 60 minutes (a 50% increase).

There were other issues (mostly around linking) impacting our project in EMSDK 2.0.9 through 2.0.12, but I could still compile and compare the compile portions to determine that the change happened between EMSDK 2.0.12 and EMSDK 2.0.13.

What changed between EMSDK 2.0.12 and 2.0.13?  At least Clang changed from 12 to 13, harfbuzz was introduced, and I don't know what else.

In our build, we have been using many -s's including:
-s USE_SDL=2 -s USE_SDL_TTF=2 -s USE_FREETYPE=2

We are using FreeType but not SDL and had this comment in our setup from long ago:
; Although we are avoiding SDL, we need USE_SDL_TTF=2 for ft2build.h (not sure why it isn't included as part of USE_FREETYPE=2)

The code didn't depend on -s USE_SDL=2 -s USE_SDL_TTF=2 other than that .h file, so I removed them.  Once I did that, our total build time went back to what it was before.  I suspected I might get an error for ft2build.h based on our comment, but whatever that issue was from before has since been removed/resolved, so it actually built cleanly.

Therefore, it appears that just including -s USE_SDL=2 -s USE_SDL_TTF=2 even though there wasn't code that needed it caused the build time (compile time was at least 25% of it but seemingly not all of the increased time) to greatly increase.  Why is having those 2 -s's causing such longer build times?

I have removed those 2 -s's from our project as we don't need them any longer, but I wanted to report the regression in build times caused by there inclusion and presumably the harfbuzz inclusion in case it points to a performance/setup problem with that code which is impacting others who need those 2 -s's (and who knows, we might need them in the future too and wouldn't want to have to pay this high a penalty).

**Comment by juj** [comment]
The emprofile tool should easily show what is increasing the build time, but looks like the fixes to that have not landed just yet: #13464

**Comment by sbc100** [comment]
The `-s` flags you are talking about should only effect the build time the first time you build if the libraries are not already present.

Can you tell roughly where the time is going.   It will most likely be one of:

1. Building the libraries (such as harfbuzz).  I doubt very much that this is the problem.
2. Linking time (this can be very slow, especially if you are use LTO,, are you?)
3. Compiling your code (-c)..  This is unlikely to have a large slowdown, and if it does than is most likely llvm itself that has changed, not emscripten.  We would still be intereted in tracking down any major regressions there.

**Comment by rsielken** [comment]
I'm compiling a directory of files.  After doing a "git clean -fxd", I would do a mk -r on the directory which compiles all of the .c and .cpp files in the directory.

The first file to use em++ or emcc that has the -s's would cause the cache to be populated (including harfbuzz).  In my timings, I was ignoring the time spent building the ports cache.  Also, once I had built once, subsequent runs had the ports cache already built, so I didn't have to ignore the ports cache building time.
FYI - do get all the ports for all the -s's we are using was taking around 52 seconds.

I wasn't even doing the linking in my testing.  I was only doing compilation.

I think it is just compiling using em++ and emcc that is taking the time.  The only difference between the comparison runs is the -s USE_SDL=2 -s USE_SDL_TTF=2 being there (slower) or not (faster) on the em++/emcc calls.  Everything else is the same (All EMSDK 2.0.13 so LLVM, Clang, etc. all the same).

**Comment by sbc100** [comment]
That is very odd.   Once the port have been built then compile times should not be effect in any way by `-s USE_SDL=2 -s USE_SDL_TTF=2`.    Those flag only do two things:

1. Build ports if they don't exist yet
2. Add include paths to the compiler

(2) basically has no cost, and take not time.
If the ports have already been built then (1) should also have zero cost.

Even if you the port are not built, as you say (2) would only account for an extra 52 seconds of your total built time of 40 minutes so I can't see how those flags could be related to the problem.

**Comment by juj** [comment]
The PR #13464 was just landed, if you want to run emprofile on the whole build, you should be able to generate a log:

`set EMPROFILE=1` (`export EMPROFILE=1` if on Linux/macOS)
<run your slow running build>
`emprofile`

The last command will generate a single .html file that you can observe, and/or upload here. See https://github.com/emscripten-core/emscripten/pull/13465#issuecomment-777365762 for examples on what kind of output it should produce.

**Comment by kripken** [comment]
Another option for things like this is to bisect: https://emscripten.org/docs/contributing/developers_guide.html#bisecting

@sbc100 did you change land in the emsdk to allow installing a random hash? If so we should update those docs to use that feature

**Comment by rsielken** [comment]
@juj I'm trying to get the emprofile output and am running into a few issues that are presumably related to me not having used it before.

I am on Linux, so I exported EMPROFILE=1 and ran the build.  I don't have an emprofile.  I found /local/emsdk/upstream/emscripten/tools/emprofile.py, so I tried to run that which gives me errors around no options.
I tried "/local/emsdk/upstream/emscripten/tools/emprofile.py --graph --outfile=full.html" and that gave me
```
builder@Sandbox:~/sandboxes/notes/desk$ /local/emsdk/upstream/emscripten/tools/emprofile.py --graph --outfile=full.html
No profiler logs were found in path "/tmp/emscripten_toolchain_profiler_logs". Try setting the environment variable EM_PROFILE_TOOLCHAIN=1 and run some emcc commands, and then rerun "python emprofile.py --graph" again.
```
I set EM_PROFILE_TOOLCHAIN=1 and tried again.

I now get a full.html and full.json file, but when I look at full.html in a browser and set the field to zero, I still don't see anything graphed.

I am running EMSDK in a Docker container, so I don't have a UI with a browser inside Docker.  I have the files outside the Docker.  I assume all I need is full.html and full.json, but I tried copying out /tmp/emscripten_toolchain_profiler_logs and that didn't help either.  Full.html should load full.json but I don't have any data showing up when I hit Refresh even with zero which should show all.

The full.json file appears to have real data in it (I only compiled a few files and not the whole directory in this recreation as I was just trying to get the tool to work and didn't want to waste all the time building the full directory just to have the tool not work).
[Full.zip](https://github.com/emscripten-core/emscripten/files/6151972/Full.zip)

What am I doing wrong/missing?

**Comment by rsielken** [comment]
I did want to add one thing that I noticed while using em++ with -v:

`
builder@Sandbox:~/sandboxes/notes/desk$ em++ -DEMSDK -DMOBILE -DOPENGL -DNO_DYNAMIC_LOADING -DWASM_NSL -g2 -s USE_PTHREADS=1 -s PROXY_TO_PTHREAD=1 -s PTHREAD_POOL_SIZE=8 -s FETCH=1 -s FETCH_SUPPORT_INDEXEDDB=0 -s FULL_ES2=1 -s USE_SDL=2 -s USE_SDL_TTF=2 -s USE_FREETYPE=2 -s USE_LIBPNG=1 -s DISABLE_EXCEPTION_CATCHING=0 -s INITIAL_MEMORY=268435456 -s ALLOW_MEMORY_GROWTH=1 -s FORCE_FILESYSTEM=1 -s ERROR_ON_UNDEFINED_SYMBOLS=0 -s OFFSCREEN_FRAMEBUFFER=1 -s DEMANGLE_SUPPORT=1 -s ABORT_ON_WASM_EXCEPTIONS=1  -c -O0 -w  -m32 -march=pentium3 -DGCC3 -DGCC4 -fno-strict-aliasing -DGCC_LBLB_NOT_SUPPORTED  -DUNIX -DLINUX -DLINUX86 -DW -DW32 -DMARVELCLIENT_INTEGRATION -DEMSDK -DWSS_CLIENT -DWSS_CLIENT_NET -fcheck-new -DPTHREAD_KERNEL -D_REENTRANT -DUSE_THREADSAFE_INTERFACES -D_POSIX_THREAD_SAFE_FUNCTIONS -DHANDLE_IS_32BITS -DHAS_IOCP -DHAS_BOOL -DHAS_DLOPEN -DUSE_PTHREAD_INTERFACES -DLARGE64_FILES -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE    -Isecui -Ireplui -I/vhome/builder/sandboxes/notes/xml/xmlproc -I/vhome/builder/sandboxes/notes/inc -I. -I/vhome/builder/sandboxes/notes/unix/cham -I/vhome/builder/sandboxes/notes/inc -o ConfigSettingsRequest.wasm.o -DSEG_CPLUS_RES_TEXT -v ConfigSettingsRequest.cpp
 "/local/emsdk/upstream/bin/clang++" -D__EMSCRIPTEN_PTHREADS__=1 -DEMSCRIPTEN -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -I/local/emsdk/upstream/emscripten/cache/sysroot/include/freetype2/freetype -I/local/emsdk/upstream/emscripten/cache/ports-builds/harfbuzz/include/harfbuzz -Xclang -isystem/local/emsdk/upstream/emscripten/cache/sysroot/include/SDL2 -DTTF_USE_HARFBUZZ=1 -target wasm32-unknown-emscripten -D__EMSCRIPTEN_major__=2 -D__EMSCRIPTEN_minor__=0 -D__EMSCRIPTEN_tiny__=15 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -flegacy-pass-manager -Werror=implicit-function-declaration --sysroot=/local/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/compat -DEMSDK -DMOBILE -DOPENGL -DNO_DYNAMIC_LOADING -DWASM_NSL -c -I/usr/local/include/g++ -I/usr/include/g++ -O0 -w -m32 -march=pentium3 -DGCC3 -DGCC4 -fno-strict-aliasing -DGCC_LBLB_NOT_SUPPORTED -DUNIX -DLINUX -DLINUX86 -DW -DW32 -DMARVELCLIENT_INTEGRATION -DEMSDK -DWSS_CLIENT -DWSS_CLIENT_NET -fcheck-new -DPTHREAD_KERNEL -D_REENTRANT -DUSE_THREADSAFE_INTERFACES -D_POSIX_THREAD_SAFE_FUNCTIONS -DHANDLE_IS_32BITS -DHAS_IOCP -DHAS_BOOL -DHAS_DLOPEN -DUSE_PTHREAD_INTERFACES -DLARGE64_FILES -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -Isecui -Ireplui -I/vhome/builder/sandboxes/notes/xml/xmlproc -I/vhome/builder/sandboxes/notes/inc -I. -I/vhome/builder/sandboxes/notes/unix/cham -I/vhome/builder/sandboxes/notes/inc -DSEG_CPLUS_RES_TEXT -v -pthread ConfigSettingsRequest.cpp -o ConfigSettingsRequest.wasm.o
clang version 13.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project 1c5f08312874717caf5d94729d825c32845773ec)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /local/emsdk/upstream/bin
 (in-process)
 "/local/emsdk/upstream/bin/clang-13" -cc1 -triple wasm32-unknown-emscripten -emit-obj -mrelax-all --mrelax-relocations -disable-free -main-file-name ConfigSettingsRequest.cpp -mrelocation-model static -mframe-pointer=none -relaxed-aliasing -fno-rounding-math -mconstructor-aliases -target-feature +atomics -target-feature +bulk-memory -target-feature +mutable-globals -target-feature +sign-ext -target-cpu generic -fvisibility hidden -debugger-tuning=gdb -v -fcoverage-compilation-dir=/vhome/builder/sandboxes/notes/desk -resource-dir /local/emsdk/upstream/lib/clang/13.0.0 -D __EMSCRIPTEN_PTHREADS__=1 -D EMSCRIPTEN -I /local/emsdk/upstream/emscripten/cache/sysroot/include/freetype2/freetype -I /local/emsdk/upstream/emscripten/cache/ports-builds/harfbuzz/include/harfbuzz -D TTF_USE_HARFBUZZ=1 -D __EMSCRIPTEN_major__=2 -D __EMSCRIPTEN_minor__=0 -D __EMSCRIPTEN_tiny__=15 -D _LIBCPP_ABI_VERSION=2 -D unix -D __unix -D __unix__ -D EMSDK -D MOBILE -D OPENGL -D NO_DYNAMIC_LOADING -D WASM_NSL -I /usr/local/include/g++ -I /usr/include/g++ -D GCC3 -D GCC4 -D GCC_LBLB_NOT_SUPPORTED -D UNIX -D LINUX -D LINUX86 -D W -D W32 -D MARVELCLIENT_INTEGRATION -D EMSDK -D WSS_CLIENT -D WSS_CLIENT_NET -D PTHREAD_KERNEL -D _REENTRANT -D USE_THREADSAFE_INTERFACES -D _POSIX_THREAD_SAFE_FUNCTIONS -D HANDLE_IS_32BITS -D HAS_IOCP -D HAS_BOOL -D HAS_DLOPEN -D USE_PTHREAD_INTERFACES -D LARGE64_FILES -D _LARGEFILE_SOURCE -D _LARGEFILE64_SOURCE -I secui -I replui -I /vhome/builder/sandboxes/notes/xml/xmlproc -I /vhome/builder/sandboxes/notes/inc -I . -I /vhome/builder/sandboxes/notes/unix/cham -I /vhome/builder/sandboxes/notes/inc -D SEG_CPLUS_RES_TEXT -isysroot /local/emsdk/upstream/emscripten/cache/sysroot -internal-isystem /local/emsdk/upstream/emscripten/cache/sysroot/include/wasm32-emscripten/c++/v1 -internal-isystem /local/emsdk/upstream/emscripten/cache/sysroot/include/c++/v1 -internal-isystem /local/emsdk/upstream/lib/clang/13.0.0/include -internal-isystem /local/emsdk/upstream/emscripten/cache/sysroot/include/wasm32-emscripten -internal-isystem /local/emsdk/upstream/emscripten/cache/sysroot/include -O0 -Werror=implicit-function-declaration -w -fdeprecated-macro -fdebug-compilation-dir=/vhome/builder/sandboxes/notes/desk -ferror-limit 19 -pthread -fgnuc-version=4.2.1 -flegacy-pass-manager -fcxx-exceptions -fexceptions -fcolor-diagnostics -isystem/local/emsdk/upstream/emscripten/cache/sysroot/include/SDL2 -iwithsysroot/include/compat -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-cxx-exceptions -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -o ConfigSettingsRequest.wasm.o -x c++ ConfigSettingsRequest.cpp
clang -cc1 version 13.0.0 based upon LLVM 13.0.0git default target x86_64-unknown-linux-gnu
ignoring nonexistent directory "/local/emsdk/upstream/emscripten/cache/sysroot/include/wasm32-emscripten/c++/v1"
ignoring nonexistent directory "/local/emsdk/upstream/emscripten/cache/sysroot/include/wasm32-emscripten"
ignoring duplicate directory "/vhome/builder/sandboxes/notes/inc"
#include "..." search starts here:
#include <...> search starts here:
 /local/emsdk/upstream/emscripten/cache/sysroot/include/freetype2/freetype
 /local/emsdk/upstream/emscripten/cache/ports-builds/harfbuzz/include/harfbuzz
 secui
 replui
 /vhome/builder/sandboxes/notes/xml/xmlproc
 /vhome/builder/sandboxes/notes/inc
 .
 /vhome/builder/sandboxes/notes/unix/cham
 /local/emsdk/upstream/emscripten/cache/sysroot/include/SDL2
 /local/emsdk/upstream/emscripten/cache/sysroot/include/compat
 /local/emsdk/upstream/emscripten/cache/sysroot/include/c++/v1
 /local/emsdk/upstream/lib/clang/13.0.0/include
 /local/emsdk/upstream/emscripten/cache/sysroot/include
End of search list.
`

Why is the FreeType2 include 
 /local/emsdk/upstream/emscripten/cache/sysroot/include/freetype2/freetype
but the harfbuzz include
 /local/emsdk/upstream/emscripten/cache/ports-builds/harfbuzz/include/harfbuzz
?

The harfbuzz h files are also in
/local/emsdk/upstream/emscripten/cache/sysroot/include/harfbuzz

Could that result in a lookup difference as far as what file paths are cached or something like that?

Again, I'm not sure if this is an issue but it was something I noticed and wanted to at least mention in case it is important.

**Comment by sbc100** [comment]
That does looks like bug in the harfbuzz port, but I don't see how it would effect build times.

**Comment by juj** [comment]
@rsielken you did it all proper. The version you were on was not the latest Emscripten branch, the emprofile PR just landed, so had to use the old syntax. There are two quirks I see with Full.zip:

1) you'll have to host an ad hoc web server to view the resulting log .html file. `python -m http.server` or `emrun --no_browser .` should work. The PR #13464 fixes this by bundling the generated .json file into the .html file.
2) for some reason the d3.min.js did not want to load even then, so I copied it to the same directory as the generated log file. (and edited the log .html file for `<script src="https://d3js.org/d3.v3.min.js">` -> `<script src="d3.v3.min.js">`.

See here for the results: http://clb.confined.space/dump/profile/ . I probably need to author another PR to fix that, not sure why that stopped working all of a sudden whereas it did work before.

There are two sources of time consumed:
 1) actual `clang++` executable takes up about ~55-60% of total execution time,
 2) remaining ~40-45% of time is actually untracked in the toolchain profiler. :( To figure out what that is, we should dig it down by enclosing the source via new `with ToolchainProfiler.profile_block('xxxx'):` blocks in the compiler. 

Would you be able to update to the latest `main` branch via git and re-do a profile? That might have better profiling blocks.

Alternatively, if you are proficient in Python, you can try adding custom profiling blocks in `emcc.py` via statements `with ToolchainProfiler.profile_block('foo'):` and then indenting the code to profile inside those blocks.

The interesting part here is that some code in `compile inputs` block is taking a long time, but when I try to reproduce the profile locally, I don't get that same performance shape for the profile at all, but it looks like this for me: http://clb.confined.space/dump/profile/foo.html?hideBlocksSmallerThan=0 . (well, this is building hello_world.c, so something is certainly being missed). Here the `compile inputs` block takes exactly the same time as the `clang++.exe` child execution.

Or is there any way to locally repro the build?

(btw, you are passing `-m32 -march=pentium3` to emcc, those don't have a meaning for wasm - although will probably be ignored anyway)

**Comment by rsielken** [comment]
@juj 

I was on tot instead of latest and thought that would do what we needed, but apparently I missed something.

Here is what I had done (to make sure I hadn't missed something, I did delete the clone, but that isn't something I normally do - I normally just do git pull):
```
git clone https://github.com/emscripten-core/emsdk.git (this leaves me with the master branch and I don't see a main branch in this repo).
./emsdk install tot
./emsdk activate tot
. ./emsdk_env.sh --build=Release
```

But, I can see that you want to be on the main branch which I found in https://github.com/emscripten-core/emscripten, so I need to do _git clone https://github.com/emscripten-core/emscripten.git_ instead/too (which has the **main** branch).

Of course, those aren't the same files, so I wasn't exactly sure what to do with them.  Looking around, I see that they appear to be what is in ./upstream/emscripten, so I just replaced the files in that directory with these.

Therefore, I have tot with /upstream/emscripten deleted and replaced.  Is that correct/valid?

That fails during the cache building of harfbuzz:
```
[ 72%] Building CXX object CMakeFiles/harfbuzz.dir/src/hb-ucdn.cc.o
[ 74%] Building C object CMakeFiles/harfbuzz.dir/src/hb-ucdn/ucdn.c.o
[ 76%] Linking C static library libharfbuzz.a
[ 76%] Built target harfbuzz
Scanning dependencies of target hb-ot-tag
[ 78%] Building CXX object CMakeFiles/hb-ot-tag.dir/src/hb-ot-tag.cc.o
[ 80%] Linking CXX executable hb-ot-tag.js
internal/modules/cjs/loader.js:883
  throw err;
  ^

Error: Cannot find module 'acorn'
Require stack:
- /local/emsdk/upstream/emscripten/tools/acorn-optimizer.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js:1:13)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/local/emsdk/upstream/emscripten/tools/acorn-optimizer.js' ]
}
em++: error: '/local/emsdk/node/14.15.5_64bit/bin/node /local/emsdk/upstream/emscripten/tools/acorn-optimizer.js /tmp/emscripten_temp_gi4bbj37/hb-ot-tag.js AJSDCE minifyWhitespace' failed (1)
```

So, instead of deleting /upstream/emscripten, I just replaced the files from the **main** branch (leaving any files that aren't in the **main** branch) so that I wouldn't be missing files.

That worked and built all the caches and my files.  I cleaned the profile history and then ran twice (once with -s USE_SDL=2 -s USE_SDL_TTF=2 and once without) and got profile reports for each (attached here).  You can see the differences in the two reports.

Now that I think I have this running correctly on **main**, I can try other things.  However, I'm not proficient in Python, but if you wanted to give detailed instructions or attach a particular file, I could replace it and try with it.

I have not tried to extract these parts out of the larger project, replace all the confidential parts, and recreate the problem as I was hoping that something simple or that could be gained by the profiler would be sufficient.  I can attempt to go down that path if needed, but I'm afraid of how hard it might be to recreate given that you were unable to recreate the same performance shape.

I will clean up the -m32 and -march flags (they are holdovers from other compilers that we inherited and didn't both to completely clean yet).

**Comment by juj** [comment]
> Therefore, I have tot with /upstream/emscripten deleted and replaced. Is that correct/valid?

The `tot` target is maintained by Google, I am not actually familiar with how that works. They way I bootstrap to git repos is with

```
./emsdk install sdk-upstream-master-64bit
./emsdk activate sdk-upstream-master-64bit
source ./emsdk_env.sh
cd emscripten/main
emcc ...
```

> got profile reports for each (attached here).

Looks like the files did not get attached?

> if you wanted to give detailed instructions or attach a particular file, I could replace it and try with it.

Basically what I'd be interested in debugging is where that large unknown time inside the `compile inputs` section from the earlier profile is going towards. If you edit emcc.py, you'll see around linie 1930 something like
```py
  with ToolchainProfiler.profile_block('compile inputs'):
    def is_link_flag(flag):
      if flag.startswith('-nostdlib'):
        return True
      return flag.startswith(('-l', '-L', '-Wl,'))
    
    # .. more code..
```
until line 2060:
```py
  # exit block 'compile inputs'
  log_time('compile inputs')
```
See that line 2060 has shorter indentation than the indented block (2 spaces vs 4 spaces). They way to measure execution time with any python code with emprofile is to add a new profile block, e.g. if you had

```py
  some_code()
  ....
  some_more_code()
  ....
  even_some_more_code()
```

and you transformed it into

```py
  with ToolchainProfiler.profile_block('my profile block'):
    some_code()
    ....
    some_more_code()

  even_some_more_code()
```
that would measure and display how much time the first two functions `some_code()` and `some_more_code()` takes.

In this case, e.g. I wonder what the profile would show if you applied the attached diff

[add_profiling.zip](https://github.com/emscripten-core/emscripten/files/6157672/add_profiling.zip)

You can also find the same in the latest commit in this branch: https://github.com/juj/emscripten/tree/profile_compile_inputs, i.e. https://github.com/juj/emscripten/commit/af01af65c0d53bd3998a801e0e0df67a9508b72c .

That should make the compile inputs section more fine grained, and help figure out a bit more what it is spending time on.

**Comment by rsielken** [comment]
Profile with emcc.py debugging added:
[emprofile.20210317_1929_AdditionalDebug.html.zip](https://github.com/emscripten-core/emscripten/files/6159295/emprofile.20210317_1929_AdditionalDebug.html.zip)

**Comment by juj** [comment]
Thanks! I think I found the root cause now. Submitted a PR https://github.com/emscripten-core/emscripten/pull/13694 to improve emprofile to capture that source. If you'd like to check out that branch to confirm it in your emprofile, that'd be great.

It looks like the issue is with Emscripten checking which tagged version of the port is installed on the local system. That involves unzipping a .tar.bz2 file which is the slow operation.

**Comment by sbc100** [comment]
> Thanks! I think I found the root cause now. Submitted a PR #13694 to improve emprofile to capture that source. If you'd like to check out that branch to confirm it in your emprofile, that'd be great.
> 
> It looks like the issue is with Emscripten checking which tagged version of the port is installed on the local system. That involves unzipping a .tar.bz2 file which is the slow operation.

OGM.. that seems really bad.. does that mean we are unzipping all the zip files of all the ports we use for every source file? 
Thank god for profiling!

**Comment by rsielken** [comment]
I cloned https://github.com/juj/emscripten/tree/profile_ports_fetch and checked out bb6e99df47495d4ae5722315e05c6cea142d336e.  I then updated the files and ran as before.  I see the delays and they are correctly identified as expected:

![image](https://user-images.githubusercontent.com/61422780/111680541-b5670980-87f8-11eb-9297-22c88dfa2b5e.png)

[emprofile.20210318_1842.html.zip](https://github.com/emscripten-core/emscripten/files/6166545/emprofile.20210318_1842.html.zip)

I assume you opened a related issue for the slow .tar.bz2 operation, but maybe you are just keeping this issue for that.  If this issue isn't for the .tar.bz2 operation, can you please link to what it now tracking that?

**Comment by sbc100** [comment]
Yes, hopefully https://github.com/emscripten-core/emscripten/pull/13702 should fully address this issue.

Looks like .tar.bz2 and .tar.gz file are fundamentally slow to list contents: https://superuser.com/questions/712095/list-contents-of-large-tar-archive-quickly

**Comment by sbc100** [comment]
Fix landed in #13702 which should be part of 2.0.16 when its released and is already available in the "tot" builds.

Please re-open if this doesn't remove all your slowdown.
